File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.cpp
class Solution {
public:
    long long gcdSum(vector<int>& nums) {
        int n = nums.size();
        vector<long long>prefix(n);
        long long maxi = 0;
        for(int i=0;i<n;i++){
            maxi = max(maxi,(long long)nums[i]);
            prefix[i] = gcd((long long)nums[i],maxi);
        }
        sort(prefix.begin(),prefix.end());
        int l = 0,r = n-1;
        long long ans = 0;
        while(l<r){
            ans+=gcd(prefix[l],prefix[r]);
            l++;
            r--;
        }
        return ans;
    }
};

File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.py
from math import gcd

class Solution:
    def gcdSum(self, nums):
        n = len(nums)

        prefixGcd = [0] * n
        mx = 0

        for i, x in enumerate(nums):
            mx = max(mx, x)
            prefixGcd[i] = gcd(x, mx)

        prefixGcd.sort()

        ans = 0

        for i in range(n // 2):
            ans += gcd(prefixGcd[i], prefixGcd[n - 1 - i])

        return ans      